### ISSUE TITLE
JavaScript calling .NET Async Method with Return Value on HostObject fail to return correct Value

### ISSUE THREAD

**Issue by RickStrahl** [MicrosoftEdge/WebView2Feedback#822]
I'm trying to call back from JavaScript into .NET using a Host object registered into the document. In experimenting I'm finding inonsistent behavior with the object registered.

For my test class I have:

```cs
  [ClassInterface(ClassInterfaceType.AutoDual)]
    [ComVisible(true)]
    public class DotnetInterop
    {

        // asyn just because
        public async Task<string> HelloWorld(string name)
        {
                   await Task.Delay(200);
                   var res = $"Hello World, {name}! - Message: 'This text was retrieved from .NET and shown here in JavaScript.'\n" +
                      "Button click in .NET -> JavaScript Function calls Hello World in .NET -> Returns message to JavaScript";
            return res;
        }
        ...
}
```

Then during init:

```cs
webView.CoreWebView2.AddHostObjectToScript("mm", Interop);
```

Now when I call this from JavaScript like this:

```js
async function callHelloWorldDotnet(name) {
    // .NET object reference (async)
    var msg = await window.chrome.webview.hostObjects.mm.HelloWorld(name);
    alert(msg);
}
```

I get a 'null' result. Also tried a promise directly. The .NET method gets called and returns a result, but the client is not picking it up **as an async function** or via a promise. If I return a promise the promise says `pending`.

Now I can get this to work by changing the method to a non-async method:

```cs
public string HelloWorld(string name)
{
    var res = $"Hello World, {name}! - Message: 'This text was retrieved from .NET and shown here in JavaScript.'\n" +
              "Button click in .NET -> JavaScript Function calls Hello World in .NET -> Returns message to JavaScript";
    return res;
}
```

Now using that same client side code works. It also works with sync from the client. **Both of these work** (in their respective async/non-async function contexts):

```js
// sync
var msg = window.chrome.webview.hostObjects.mm.sync.HelloWorld(name);

// async
var msg = await window.chrome.webview.hostObjects.mm.HelloWorld(name);
```

Summary: 

* I can call the method both sync and async from JavaScript if .NET Method is **not async**
* I can call the method both sync and async from JavaScript if the .NET is async **and** no value is returned (fire and forget)
* I cannot call the method **either** async or sync from JavaScript if .NET method is async and returns a value

### Question: When do we need Async .NET Host Methods?
My question is when the client is calling the host using the async instance shouldn't the method calls on the host object be async as well or at least be supported as async? 

All of this feels very inconsistent, but it's not clear if this is by design or a pre-release bug...

In my actual application for now I've been avoiding async calls for results that return values, because I needed it to work both with WV2 and WebBrowser so I use sync. However, in my test form which is green field I can use aysnc through all of it and expected the called methods also to be async. 

Note it looks to me that the methods fired are getting invoked as async methods - meaning the code in the async methods executes as it should with async semantics. But the results are not picked up as TASK based async values and returned as is rather than fixing up the async call.

Either way a clear answer on how this is supposed to work would be helpful.

### Desired/Expected behavior:

* If using the `.sync` object from JS call with sync semantics
* If not use the `.sync` object from JS call with async/await semantics both on client and host

**Comment by champnic** [comment]
Thanks for the detailed summary @RickStrahl. This is a known limitation that async .NET functions aren't currently supported. We are tracking support in #75, but there's also a potential workaround here (if changing the .NET functions to sync doesn't work for you):
https://github.com/MicrosoftEdge/WebView2Feedback/issues/75#issuecomment-708144507

**Comment by RickStrahl** [comment]
So what is the point of having an `async` object if it doesn't actually work async?

**Comment by champnic** [comment]
Do you mean the JS projection, `window.chrome.webview.hostObjects.<object>`? That's currently async by default as it's a cross-process call back to the host app.

If you mean the .NET host object, there's currently no point in it being async until we add support.

**Comment by RickStrahl** [comment]
Yes that's what I'm talking about. 

The problem is that you can't return a value from an async call. The call is async, but you can't actually await it on the JavaScript side because the result is a pending promise that never resolves (unless this was fixed).

It runs async, but if you want a result back (which is why you'd need async in the first place) then you're out of luck.

I don't understand how the sync calls can work and async won't. If the calls are out of band (they have to be because OOP) then code has to wait for completion no matter what. I don't see how this can not be wired up into a working promise interface since it works for the sync call?

On my end this bug is causing a lot of problems because it breaks the async chain in .NET.  The only way I can handle a .NET method that returns a value I have to make the method sync or it won't work to return that value. If I now need to make an async call inside of that called code I don't have a way to do this cleanly. This comes up a lot in my use case and because of it I fight with deadlocks.

Bottom line - I think this would be a worthwhile thing to fix. For behavior like I describe, but also just for pure consistency. Having an interface that works async but fails to return a proper async result is just broken.

**Comment by champnic** [comment]
Ok I think we're in agreement. If the .NET host object has an async function that's getting called, that's what's currently broken (and as you mentioned, has dangerous workarounds). That is the support we are planning on adding in #75 - I've updated the title to make it more clear. That issue has been open for a long time, so I'll see if I can bump the priority and get a better sense from the team if it's something that wouldn't be too expensive to implement. Thanks!

**Comment by RickStrahl** [comment]
@champnic Finally coming back to this to see if I can make async calls work and I don't seem to be able to call a method async without problems.

Same code works sync, but asycn fails.

Here's what I'm doing:

```javascript
dotnetInterop: {
    // return async instance if available, otherwise sync
    getEditor: function() {  // async instance
      if (window.chrome) {
        // WebView instance does not want to be cached
        return window.chrome.webview.hostObjects.mm || window.chrome.webview.hostObjects.sync.mm;
      }
      return te.mmEditorAsync || te.mmEditor || null;
    },
    getEditorSync: function() {
      if (window.chrome) {
        // WebView instance does not want to be cached
        return window.chrome.webview.hostObjects.sync.mm;
      }
      return te.mmEditor || null;
    },
    // IMPORTANT: Method is async!
    previewLinkNavigation: async function(url, rawHref) {
        //var editor = te.dotnetInterop.getEditorSync();   //  this works if making a sync call
        var editor = te.dotnetInterop.getEditor();
      if (!editor)
        return false;
      try {
        return await editor.PreviewLinkNavigation(url, rawHref);
      } catch (ex) { return false; }
    },
```

Then the code that uses this block:

```javascript
$(document).on("click",
        "a",
        async (e) => {
            var url = this.href;
            var rawHref = $(this).attr("href");
            var hash = this.hash;

            if (te.mmEditor) {
                debugger;
                var handled = await te.dotnetInterop.previewLinkNavigation(url, rawHref);   // blows up async - works sync
                if (handled) {
                    // it true editor handled the navigation
                    e.preventDefault();
                    return false;
                }
            }
...
        });
```

The .NET code gets called and it completes properly from the .NET end, but when the call returns the script fails and exits without running any further code (unhandled exception).

The .NET code called works but just for reference here is the signature:

```cs
public async Task<bool> PreviewLinkNavigation(string url, string src = null)
...

await RaiseOnPreviewLinkNavigation(url, src);

...
```

and I'm passing both parameters in.

As many months ago the async call does not seem to work for me.

**Comment by RickStrahl** [comment]
So I gave this a try with the latest preview runtime component (1.0.1189-prerelease)  but **not** the newer matching runtime, and I still cannot get an async call to return a value.

Here's what I'm doing and seeing.

The following code is called on a click handler to pass through links from the browser to .NET to examine and handle so they open in an external browser or in the appropriate applications or editors.

```javascript
  dotnetInterop: {
    // return async instance if available, otherwise sync
    getEditor: function() {
      if (window.chrome) {
        // WebView instance does not want to be cached
        return window.chrome.webview.hostObjects.mm || window.chrome.webview.hostObjects.sync.mm;
      }
      return te.mmEditorAsync || te.mmEditor || null;
    },

// ... other methods

    previewLinkNavigationAsync: async function (url, rawHref) {
        var editor = te.dotnetInterop.getEditor();  // get async .NET editor instance
        if (!editor)
            return false;

        debugger;
        try {
            // **** THIS FAILS HERE! Hard fault, `catch` does not fire
            var handled = await editor.PreviewLinkNavigationAsync(url, rawHref);  
 

           return handled;  <--- never gets called
        } catch (ex) {  //   <---- this doesn't fire on the failure
            debugger;
            console.log(ex);
        }
    },
}
```


The . NET async method code that is called gets fired just fine: I can see it running in the debugger and it's correctly receiving the iinput and correctly returning a value:

```c#
public async Task<bool> PreviewLinkNavigationAsync(string url, string src = null)
{
    var editor = Model.ActiveEditor;
    bool result =  await editor.PreviewLinkNavigation(url, src);
    return result;
}
```

also tried something simpler to make sure there's not some problem with the code called in the background (which uses Shell to open a link):

```cs
public async Task<bool> PreviewLinkNavigationAsync(string url, string src = null)
{
    var editor = Model.ActiveEditor;
    await Task.Yield();
    return true;
}
```

with the same result - the code simply fails silently without firing the catch handler.


![asyncWebViewFail](https://user-images.githubusercontent.com/1374013/160064081-4b244bbb-bdd4-4664-ba89-c88f9210a75f.gif)

You can see the code firing into this method and run up to the `await` call - stepping F10 over the line then jumps out of the debugger altogether and the external page is opened and navigated (from WPF shell navigation which in my case opens Brave).

**Comment by RickStrahl** [comment]
Just to make sure there's isn't something funky happening with the async/await support I also set up a raw Promise call:

```js
    previewLinkNavigationAsync: function (url, rawHref) {
        var editor = te.dotnetInterop.getEditor();  // get async .NET editor instance
        if (!editor)
            return false;

        try {
            var handled = editor.PreviewLinkNavigationAsync(url, rawHref);
            console.log(handled);   <!-- promise
            handled
                .then((val) => {
                    console.log('then: ' + val);   <!-- fires but always `null`
                })
                .catch((err) => {
                    console.log(err);
                });
            return handled;
        } catch (ex) {
            debugger;
            console.log(ex);
        }
    },
```

When I do this `then` fires, but the result value is always `null`.

So in summary this is exactly the same behavior I was describing above - I don't see that anything has changed.

**Comment by RickStrahl** [comment]
Just to clarify some of the scenario I've tried and what does and doesn't work:

**works**

* Sync JS call to .NET method - method gets called result works (off of sync host object) - This is useful and really glad this works!
* Async JS call to .NET **without result** - .NET code gets called this work

** doesn't work**

* Async JS call to Async .NET method  with result - .NET code gets called, but the result is not returned. await calls fail hard in WebView (no error trapping including try/catch not firing).
* Sync JS call to Async .NET method - .NET method is called, but result is null (should be a promise?)

To summarize the problem really appears to be that the WebView2 runtime is not dealing with Task results correctly and failing to translate the Task results back into JavaScript promises.

**Comment by champnic** [comment]
Hey @RickStrahl - Can you try this with the newer runtime? I just checked the code and I believe it also has changes in the runtime.

**Comment by RickStrahl** [comment]
This is still not working for me. I'm running:

* 1150 Runtime (Canary?)
* 1189 WebView .NET Component

Where can I get a newer runtime than Canary?


Another thing I noticed as to what's actually happening is that the `await` call returns immediately - it doesn't wait for completion.

In this code:

```js
previewLinkNavigationAsync: async function (url, rawHref) {
        // HACK: Must call sync for result - WPF code uses async so potential lockup scenario
        var editor = te.dotnetInterop.getEditor();
        if (!editor)
            return false;
        try {
            // TODO: ASYNC RESULT DOESN'T WORK IN WEBVIEW - CHECK IN LATER VERSIONS
            var handled = await editor.PreviewLinkNavigationAsync(url, rawHref);
            alert('Handled: ' + handled);
            return handled;
        } catch (ex) { return false; }
    },
```

I see the alert box before the code in .NET is done executing with `handled` = `null`.

**Comment by champnic** [comment]
You'll need runtime 100.0.1183+. Current Canary should be 101.0.1204.0. Can you go to edge://settings/help in your Canary browser and see if there's an update available?

**Comment by RickStrahl** [comment]
My Canary version is: 101.0.1204.0

But when I run the `versionNo = CoreWebView2Environment.GetAvailableBrowserVersionString();` gives me the `1150` runtime which is my installed WebView Runtime. Should I uninstall that?

Can't though:

![image](https://user-images.githubusercontent.com/1374013/160223732-e2ea6da8-99a9-4efd-b93c-1761a1044060.png)

**Comment by champnic** [comment]
You can tell the WebView2 control to pickup the Canary channel instead:
https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/set-preview-channel

**Comment by RickStrahl** [comment]
I was unable to make that work with instructions from [here](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/set-preview-channel#win32c).

![image](https://user-images.githubusercontent.com/1374013/160229670-ef86f67b-208e-4cff-8cff-74a241e52349.png)
